File: ts-api/examples/smoke_v5.ts
/**
 * V5 smoke test against the live API (small sizes to stay in the Opus free tier).
 * Usage: pnpm exec tsx examples/smoke_v5.ts [step...]
 */
import * as fs from "fs";
import * as path from "path";
import sharp from "sharp";
import { NovelAIClient } from "../src/client";
import { summarizeOpusUsage } from "../src/anlas";

const OUT = process.env.SMOKE_OUT ?? "/tmp/novelai_smoke/ts_v5";
fs.mkdirSync(OUT, { recursive: true });
const client = new NovelAIClient();
const steps = process.argv.slice(2);
const want = (s: string) => steps.length === 0 || steps.includes(s);
const base = { model: "nai-diffusion-5-full", width: 512, height: 768, steps: 23, prompt: "1girl, solo, red apple in hand, simple background" } as const;

async function step(name: string, fn: () => Promise<string>) {
  if (!want(name)) return;
  const t0 = Date.now();
  try {
    const info = await fn();
    console.log(`OK   ${name} (${((Date.now() - t0) / 1000).toFixed(1)}s) ${info}`);
  } catch (e: any) {
    console.log(`FAIL ${name} (${((Date.now() - t0) / 1000).toFixed(1)}s) ${e?.constructor?.name}: ${String(e?.message ?? e).slice(0, 300)}`);
  }
}

async function alphaRatio(file: string): Promise<string> {
  const { data, info } = await sharp(file).ensureAlpha().raw().toBuffer({ resolveWithObject: true });
  let transparent = 0;
  for (let i = 3; i < data.length; i += 4) if (data[i] === 0) transparent++;
  return `${info.width}x${info.height} alpha0=${((transparent / (info.width * info.height)) * 100).toFixed(1)}%`;
}

const anlas = (r: { anlas_consumed?: number | null; anlas_remaining?: number | null }) => `anlas consumed=${r.anlas_consumed} remaining=${r.anlas_remaining}`;

(async () => {
  const src = path.join(OUT, "t2i.png");
  const mask = await sharp({ create: { width: 512, height: 768, channels: 3, background: "black" } })
    .composite([{ input: await sharp({ create: { width: 256, height: 256, channels: 3, background: "white" } }).png().toBuffer(), left: 128, top: 0 }])
    .png().toBuffer();

  await step("balance", async () => {
    const b = await client.getAnlasBalance();
    return `total=${b.total} tier=${b.tier} usage=${JSON.stringify(b.usage)} summary=${b.usage ? JSON.stringify(summarizeOpusUsage(b.usage)) : "-"}`;
  });
  await step("t2i_transparent", async () => {
    const r = await client.generate({ ...base, seed: 1, transparent_background: true, save_path: src });
    return `${await alphaRatio(src)} ${anlas(r)}`;
  });
  await step("t2i_webp", async () => {
    const out = path.join(OUT, "webp_dir");
    const r = await client.generate({ ...base, seed: 8, transparent_background: true, image_format: "webp", save_dir: out });
    const meta = await sharp(r.saved_path!).metadata();
    return `format=${r.image_format} file=${path.basename(r.saved_path!)} ${meta.format} ${meta.width}x${meta.height} alpha=${meta.hasAlpha} ${anlas(r)}`;
  });
  await step("i2i_from_webp", async () => {
    const webpSrc = fs.readdirSync(path.join(OUT, "webp_dir")).find(f => f.endsWith(".webp"));
    if (!webpSrc) return "skipped (no webp)";
    const r = await client.generate({ ...base, seed: 9, action: "img2img", source_image: path.join(OUT, "webp_dir", webpSrc), img2img_strength: 0.5, save_path: path.join(OUT, "i2i_from_webp.png") });
    return `format=${r.image_format} ${anlas(r)}`;
  });
  await step("i2i", async () => {
    const out = path.join(OUT, "i2i.png");
    const r = await client.generate({ ...base, seed: 2, action: "img2img", source_image: src, img2img_strength: 0.6, transparent_background: true, save_path: out });
    return `${await alphaRatio(out)} ${anlas(r)}`;
  });
  await step("infill_full", async () => {
    const out = path.join(OUT, "infill_full.png");
    const r = await client.generate({ ...base, prompt: base.prompt + ", cat ears", seed: 3, action: "infill", source_image: src, mask, mask_strength: 0.7, save_path: out });
    return `${await alphaRatio(out)} ${anlas(r)}`;
  });
  await step("t2i_curated", async () => {
    const r = await client.generate({ ...base, model: "nai-diffusion-5-curated", seed: 4, save_path: path.join(OUT, "curated.png") });
    return anlas(r);
  });
  await step("infill_curated", async () => {
    const r = await client.generate({ ...base, model: "nai-diffusion-5-curated", seed: 5, action: "infill", source_image: path.join(OUT, "curated.png"), mask, mask_strength: 0.7, save_path: path.join(OUT, "infill_curated.png") });
    return anlas(r);
  });
  await step("characters", async () => {
    const r = await client.generate({
      ...base, seed: 6, prompt: "2girls, standing, simple background",
      characters: [{ prompt: "girl, red hair", center_x: 0.3, center_y: 0.5 }, { prompt: "girl, blue hair", center_x: 0.7, center_y: 0.5 }],
      save_path: path.join(OUT, "characters.png"),
    });
    return anlas(r);
  });
  await step("vibe_rejected", async () => {
    try {
      await client.generate({ ...base, vibes: [path.resolve(__dirname, "../vibes/input1.naiv4vibe")] });
      return "UNEXPECTED: accepted";
    } catch (e: any) {
      return `rejected locally: ${String(e.message).includes("Vibe Transfer is not supported")}`;
    }
  });
  await step("declutter_keep_bubbles", async () => {
    const r = await client.augmentImage({ req_type: "declutter-keep-bubbles", image: src, save_path: path.join(OUT, "declutter.png") });
    return anlas(r);
  });
})();

File: rust-api/examples/smoke_v5.rs
//! V5 smoke test against the live API (small sizes to stay in the Opus free tier).
//!
//! Run with: cargo run --example smoke_v5 [step...]

use std::future::Future;
use std::io::Cursor;
use std::time::Instant;

use image::{ImageEncoder, RgbaImage};
use novelai_api::anlas::summarize_opus_usage;
use novelai_api::client::NovelAIClient;
use novelai_api::constants::{AugmentReqType, Model};
use novelai_api::schemas::*;

const OUT: &str = "/tmp/novelai_smoke/rust_v5";
const PROMPT: &str = "1girl, solo, red apple in hand, simple background";

fn mask_png(width: u32, height: u32) -> Vec<u8> {
    let img = RgbaImage::from_fn(width, height, |x, y| {
        if (128..384).contains(&x) && y < 256 {
            image::Rgba([255, 255, 255, 255])
        } else {
            image::Rgba([0, 0, 0, 255])
        }
    });
    let mut buf = Cursor::new(Vec::new());
    image::codecs::png::PngEncoder::new(&mut buf)
        .write_image(img.as_raw(), width, height, image::ExtendedColorType::Rgba8)
        .unwrap();
    buf.into_inner()
}

fn alpha_ratio(path: &str) -> String {
    match image::open(path) {
        Ok(img) => {
            let rgba = img.to_rgba8();
            let transparent = rgba.pixels().filter(|p| p[3] == 0).count();
            format!("{}x{} alpha0={:.1}%", rgba.width(), rgba.height(), transparent as f64 * 100.0 / (rgba.width() * rgba.height()) as f64)
        }
        Err(e) => format!("(could not read: {})", e),
    }
}

fn base(prompt: &str, model: Model, seed: u64, path: &str) -> GenerateParamsBuilder {
    GenerateParams::builder(prompt)
        .model(model)
        .width(512)
        .height(768)
        .steps(23)
        .seed(seed)
        .save_path(format!("{}/{}", OUT, path))
}

fn anlas(consumed: Option<u64>, remaining: Option<u64>) -> String {
    format!("anlas consumed={:?} remaining={:?}", consumed, remaining)
}

async fn step<F, Fut>(wanted: &[String], name: &str, f: F)
where
    F: FnOnce() -> Fut,
    Fut: Future<Output = anyhow::Result<String>>,
{
    if !wanted.is_empty() && !wanted.iter().any(|w| w == name) {
        return;
    }
    let t0 = Instant::now();
    match f().await {
        Ok(info) => println!("OK   {} ({:.1}s) {}", name, t0.elapsed().as_secs_f64(), info),
        Err(e) => println!("FAIL {} ({:.1}s) {}", name, t0.elapsed().as_secs_f64(), e),
    }
}

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    dotenvy::dotenv_override().ok();
    std::fs::create_dir_all(OUT)?;
    let wanted: Vec<String> = std::env::args().skip(1).collect();
    let client = NovelAIClient::new(None, None)?;
    let src = format!("{}/t2i.png", OUT);
    let mask = mask_png(512, 768);

    step(&wanted, "balance", || async {
        let b = client.get_anlas_balance().await?;
        let summary = b.usage.as_ref().map(summarize_opus_usage);
        Ok(format!("total={} tier={} usage={:?} summary={:?}", b.total, b.tier, b.usage, summary))
    })
    .await;
    step(&wanted, "t2i_transparent", || async {
        let p = base(PROMPT, Model::NaiDiffusion5Full, 1, "t2i.png").transparent_background(true).build()?;
        let r = client.generate(&p).await?;
        Ok(format!("{} {}", alpha_ratio(&src), anlas(r.anlas_consumed, r.anlas_remaining)))
    })
    .await;
    step(&wanted, "t2i_webp", || async {
        let p = base(PROMPT, Model::NaiDiffusion5Full, 8, "unused.png")
            .transparent_background(true)
            .image_format(novelai_api::constants::OutputFormat::Webp)
            .save_dir(format!("{}/webp_dir", OUT))
            .build()?;
        let r = client.generate(&p).await?;
        let path = r.saved_path.clone().unwrap_or_default();
        Ok(format!("format={} file={} {} {}", r.image_format, path.rsplit('/').next().unwrap_or(""), alpha_ratio(&path), anlas(r.anlas_consumed, r.anlas_remaining)))
    })
    .await;
    step(&wanted, "i2i", || async {
        let p = base(PROMPT, Model::NaiDiffusion5Full, 2, "i2i.png")
            .transparent_background(true)
            .action(GenerateAction::Img2Img { source_image: ImageInput::FilePath(src.clone().into()), strength: 0.6, noise: 0.0 })
            .build()?;
        let r = client.generate(&p).await?;
        Ok(format!("{} {}", alpha_ratio(&format!("{}/i2i.png", OUT)), anlas(r.anlas_consumed, r.anlas_remaining)))
    })
    .await;
    step(&wanted, "infill_full", || async {
        let p = base(&format!("{}, cat ears", PROMPT), Model::NaiDiffusion5Full, 3, "infill_full.png")
            .action(GenerateAction::Infill {
                source_image: ImageInput::FilePath(src.clone().into()),
                mask: ImageInput::Bytes(mask.clone()),
                mask_strength: 0.7,
                color_correct: true,
                hybrid_strength: None,
                hybrid_noise: None,
            })
            .build()?;
        let r = client.generate(&p).await?;
        Ok(anlas(r.anlas_consumed, r.anlas_remaining))
    })
    .await;
    step(&wanted, "t2i_curated", || async {
        let r = client.generate(&base(PROMPT, Model::NaiDiffusion5Curated, 4, "curated.png").build()?).await?;
        Ok(anlas(r.anlas_consumed, r.anlas_remaining))
    })
    .await;
    step(&wanted, "infill_curated", || async {
        let p = base(PROMPT, Model::NaiDiffusion5Curated, 5, "infill_curated.png")
            .action(GenerateAction::Infill {
                source_image: ImageInput::FilePath(format!("{}/curated.png", OUT).into()),
                mask: ImageInput::Bytes(mask.clone()),
                mask_strength: 0.7,
                color_correct: true,
                hybrid_strength: None,
                hybrid_noise: None,
            })
            .build()?;
        let r = client.generate(&p).await?;
        Ok(anlas(r.anlas_consumed, r.anlas_remaining))
    })
    .await;
    step(&wanted, "characters", || async {
        let p = base("2girls, standing, simple background", Model::NaiDiffusion5Full, 6, "characters.png")
            .characters(vec![
                CharacterConfig { prompt: "girl, red hair".into(), center_x: 0.3, center_y: 0.5, ..Default::default() },
                CharacterConfig { prompt: "girl, blue hair".into(), center_x: 0.7, center_y: 0.5, ..Default::default() },
            ])
            .build()?;
        let r = client.generate(&p).await?;
        Ok(anlas(r.anlas_consumed, r.anlas_remaining))
    })
    .await;
    step(&wanted, "vibe_rejected", || async {
        let built = base(PROMPT, Model::NaiDiffusion5Full, 7, "never.png")
            .vibes(vec![VibeConfig { item: VibeItem::FilePath("vibes/input1.naiv4vibe".into()), strength: 0.7, info_extracted: 0.7 }])
            .build();
        Ok(match built {
            Err(e) => format!("rejected locally: {}", e),
            Ok(_) => "UNEXPECTED: accepted".into(),
        })
    })
    .await;
    step(&wanted, "declutter_keep_bubbles", || async {
        let r = client
            .augment_image(&AugmentParams {
                req_type: AugmentReqType::DeclutterKeepBubbles,
                image: ImageInput::FilePath(src.clone().into()),
                prompt: None,
                defry: None,
                save: SaveTarget::ExactPath(format!("{}/declutter.png", OUT)),
            })
            .await?;
        Ok(anlas(r.anlas_consumed, r.anlas_remaining))
    })
    .await;
    Ok(())
}
